File: 프로그래머스/1/42576. 완주하지 못한 선수/완주하지 못한 선수.py
'''
1.아이디어
    - 주어진 2개의 배열에서 서로 겹치는 인물을 제거하고 남은 participant 반환하기
        1) 동명이인이 있기 때문에 set() 사용 불가능
        1-2) 두 배열의 차집합을 구하면 되지만, 중복이 있을 경우 처리 불가능
        1-3) 일일이 비교해서 제거하기 & find()를 이용하기 > 시간초과
        >> collections.Counter를 활용하여 등장 횟수까지 저장된 자료구조 활용 > 차집합 처럼 사용
'''

from collections import Counter


def solution(participant, completion):
    
    part_cnt = Counter(participant)
    com_cnt = Counter(completion)

    # 차집합처럼 동작. 값이 0이 아닌 키만 남음
    answer = part_cnt - com_cnt
    return list(answer.keys())[0]

File: 프로그래머스/1/42576. 완주하지 못한 선수/완주하지 못한 선수.java
import java.util.*;

class Solution {
    public String solution(String[] participant, String[] completion) {
        // participants를 map 형태로 변경하기
        Map<String, Integer> map = new HashMap<>();
        for(String name: participant){
            map.put(name, map.getOrDefault(name,0)+1);
        }
        
        // completion에 존재하는 이름이면 map에서 제거하기
        for(String name: completion){
            if(map.get(name) > 0){
                map.put(name, map.get(name)-1);
            }
        }
        
        // 완주하지 못한 선수 찾기
        for (Map.Entry<String, Integer> entry : map.entrySet()) {
            if (entry.getValue() >= 1) {
                return entry.getKey();
            }
        }
        
        return "";
    }
}
